introduce some support for native gems

flavorjones opened the pull request:
Native gems were briefly discussed at #8 and I was able to do a little bit of work to add support for native (precompiled) gems.

To try this out, you'll need Docker installed on your system. Then:

* `bundle update` to install rake-compiler-dock
* `bundle exec rake clean`
* `bundle exec rake gem:x86_64-linux` to build that platform's native gem

I've also added `x86-linux` to this commit; but not Darwin because I had trouble getting libmagic to cross-compile (we can look into that later).

The packaged gem will the contain a `lib` directory that looks like:
- lib/magic/2.5/magic.so
- lib/magic/2.6/magic.so
- lib/magic/2.7/magic.so
- lib/magic/3.0/magic.so

Each of these is a dynamic shared library that contains a statically-linked version of libmagic, with the C extension compiled against that version of Ruby.

Let me know what you think, and I hope this is helpful.

flavorjones committed:
Rakefile: rename local variable to avoid Bundler collision

Bundler will define a `Kernel.gem` method which will bite us when we
try to cross-compile later when it conflicts with this local variable.

flavorjones committed:
dev: add '/pkg' to the CLEAN list

so that it's easily blown away when we start building multiple native gems

flavorjones committed:
ext: look for some needed system libraries

before trying to resolve rb_thread_call_without_gvl. Some linux
systems will fail to find symbols in these libraries (specifically,
rake-compiler-dock's linux images).

flavorjones committed:
ext: look in a ruby-version-specific subdir for the .so

because this is where rake-compiler will put the compiled library in a
native gem that supports multiple versions of ruby.

Note that we will fall back to the regular location for people who are
installing from source, either in dev or from the vanilla ruby gem.

flavorjones committed:
ext: support native precompiled gems.

I've arbitrarily picked Ruby 2.5-3.0, but rake-compiler-dock will
support back to 2.3 if that's what you want.

I haven't added Darwin platforms here because there are some
challenges building libmagic that I didn't have time to look into.

stanhu commented:
@flavorjones Thanks, this is awesome! I do see these warnings at the beginning of the build:

```
﻿﻿no configuration section for specified version of Ruby (rbconfig-x86-linux-2.5.0)
no configuration section for specified version of Ruby (rbconfig-x86-linux-2.6.0)
no configuration section for specified version of Ruby (rbconfig-x86-linux-2.7.0)
no configuration section for specified version of Ruby (rbconfig-x86-linux-3.0.0)
```

Is that expected?

After installing the gem, I also got this:

```
irb(main):001:0> require 'magic'
Traceback (most recent call last):
       10: from /Users/stanhu/.rbenv/versions/2.7.2/bin/irb:23:in `<main>'
        9: from /Users/stanhu/.rbenv/versions/2.7.2/bin/irb:23:in `load'
        8: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        7: from (irb):1
        6: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:156:in `require'
        5: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:168:in `rescue in require'
        4: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:168:in `require'
        3: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.0-x86_64-linux/lib/magic.rb:3:in `<top (required)>'
        2: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.0-x86_64-linux/lib/magic.rb:7:in `rescue in <top (required)>'
        1: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.0-x86_64-linux/lib/magic.rb:7:in `require_relative'
LoadError (cannot load such file -- /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.0-x86_64-linux/lib/magic/magic)
```

Seems we are missing the `share` files that are normally installed with the `libmagic` compilation.

flavorjones commented:
> no configuration section for specified version of Ruby

Yes, these are emitted by rake-compiler to warn about the platform/ruby-version matrix elements that **that** specific container can't build. Safe to ignore.

I can't reproduce the error you're seeing. You may want to uninstall any versions of ruby-magic that you already have installed, and then try to install the generate gem file again. Here's what I see (on x86_64):

```
$ gem install --local pkg/ruby-magic-0.4.0-x86_64-linux.gem 
Thank you for installing!
Successfully installed ruby-magic-0.4.0-x86_64-linux
1 gem installed

$ irb -rmagic
2.7.2 :001 > Magic.new.file("/bin/ls")
 => "ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=2f15ad836be3339dec0e2e6a3c637e08e48aacbd, for GNU/Linux 3.2.0, stripped" 
2.7.2 :002 > 
```

stanhu commented:
I noticed that `paths` is set to the local build directory instead of the installed path:

```ruby
irb(main):001:0> require 'magic'
=> true
irb(main):002:0> Magic.new
=> #<Magic:0x000055b4b48a42d8 @flags=0, @paths=["/home/stanhu/ruby-magic/ports/x86_64-redhat-linux/libmagic/5.39/share/misc/magic"]>
```

If you run `bundle exec rake clobber` to clean up `ports/archive`, you see:

```ruby
irb(main):001:0> require 'magic'
=> true
irb(main):002:0> Magic.new
Traceback (most recent call last):
        6: from /home/stanhu/.rbenv/versions/2.7.3/bin/irb:23:in `<main>'
        5: from /home/stanhu/.rbenv/versions/2.7.3/bin/irb:23:in `load'
        4: from /home/stanhu/.rbenv/versions/2.7.3/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        3: from (irb):2
        2: from (irb):2:in `new'
        1: from (irb):2:in `initialize'
Magic::MagicError (could not find any valid magic files!)
```

We need the `.gem` to include the `magic.mgc` file inside the `share` path.

kwilczynski commented:
Hi @stanhu,

> I noticed that `paths` is set to the local build directory instead of the installed path:
> 
> ```ruby
> irb(main):001:0> require 'magic'
> => true
> irb(main):002:0> Magic.new
> => #<Magic:0x000055b4b48a42d8 @flags=0, @paths=["/home/stanhu/ruby-magic/ports/x86_64-redhat-linux/libmagic/5.39/share/misc/magic"]>
> ```
[...]

I have on my "to do" list an item which is about looking into changing the location of the compiled Magic database, so that we would move it perhaps into the `lib/magic`, but this requires a small patch most likely, as I don't think there is a way to influence location of where the database will be placed other than setting the prefix (which would make sense if we were to then follow with `make install` or `make DESTDIR="/some/path/somewhere" install`), which we aren't really doing in this case - also it does not quite solve the problem. :)

Krzysztof

flavorjones commented:
Ah, what you're both saying makes sense to me now.

Nokogiri copies some "extra" files from libxml2 (the C header files, specifically) into the gem as well. I imagine ruby-magic could do this pretty easily, too, so long as we can resolve the path dynamically (to wherever it's installed). I'll poke around a bit this afternoon.

flavorjones committed:
dev: don't clobber tarballs

flavorjones committed:
test: add coverage for Magic#path after #load

flavorjones committed:
prefactor: always use @paths to store paths

- rb_mgc_get_paths saves default values to the ivar
- rb_mgc_load uses the ivar if no arguments are passed
- use of magic_getpath_wrapper() is limited to just one place

which will make it easy to modify the default path in an upcoming
commit

flavorjones committed:
packaging: copy magic.mgc into the native gems

Note that this commit doesn't _use_ the file, just packages it.

flavorjones committed:
feat: use the packaged magic.mgc file if it's present

Note that the behavior of Magic#paths has subtly changed here, where we'll only do a lookup the first time (either to magic_getpath_wrapper() or to Magic.default_paths), and from then on the object will use the cached value.

Previously, if the MAGIC env var was set, a lookup would happen every time.

flavorjones committed:
package: slim the native gem

- exclude tarballs and patches
- drop mini_portile2 as a dependency

flavorjones committed:
test: add coverage for how #paths and #load work with ENV['MAGIC']

flavorjones commented:
I've added some commits that will:

- package the `magic.mgc` file in native gems at `ext/magic/share/magic.mgc`
- use that database file if it exists (otherwise fall back to the current behavior)
- remove patches, tarball, and the mini_portile2 dependency from native gems
- add some test coverage for `Magic#paths` and `Magic#load` when the `MAGIC` env var is set

Let me know what you think!

flavorjones commented:
Just to clarify what happens here ...

If the `ruby` platform gem is installed, and libmagic is compiled at install time, then `Magic#paths` will be the default compiled into libmagic, something like:

```
/path/to/gems/ruby-magic-0.4.0/ports/x86_64-pc-linux-gnu/libmagic/5.39/share/misc/magic
```

If the native platform gem is installed, then `Magic#paths` will be overridden, and will be the file we install, something like:

```
/path/to/gems/ruby-magic-0.4.0-x86_64-linux/ext/magic/share/magic.mgc
```

stanhu commented on ext/magic/ruby-magic.c:
```diff
@@ -294,15 +294,19 @@ rb_mgc_get_paths(VALUE object)
-	RB_GC_GUARD(value);
-
-	return value;
+  }
```
Nit: Should this `}` be indented here?

flavorjones replied:
tabs vs spaces! :shakes-fist-at-editor: Will fix.

stanhu commented on ext/magic/ruby-magic.c:
```diff
@@ -294,15 +294,19 @@ rb_mgc_get_paths(VALUE object)
+  /* use either the library method or our gem-specific default, and save the result in @paths */
+  value = rb_funcall(rb_cMagic, rb_intern("default_paths"), 0);
+  if (getenv("MAGIC") || NIL_P(value)) {
+    cstring = magic_getpath_wrapper();
```
Nit: This indentation should probably be consistent with the above?

flavorjones replied:
tabs vs spaces! :shakes-fist-at-editor: Will fix.

stanhu commented on Rakefile:
```diff
@@ -48,15 +51,64 @@ RuboCop::RakeTask.new('lint') do |t|
+  def gem_builder(platform)
+    # use Task#invoke because the pkg/*gem task is defined at runtime
+    Rake::Task["native:#{platform}"].invoke
+    Rake::Task["pkg/#{RUBY_MAGIC_GEM_SPEC.full_name}-#{Gem::Platform.new(platform).to_s}.gem"].invoke
```
On macOS, I'm seeing:

```
/usr/local/rake-compiler/ruby/x86_64-redhat-linux/ruby-3.0.0/include/ruby-3.0.0/ruby/internal/memory.h:254:57: warning: conversion to ‘size_t’ {aka ‘long unsigned int’} from ‘long int’ may change the sign of the result [-Wsign-conversion]
  254 |         store, rbimpl_size_mul_or_raise(count, elsize), count);
      |                                                         ^~~~~
linking shared-object magic/magic.so
WARNING:  See https://guides.rubygems.org/specification-reference/ for help
rake aborted!
Gem::InvalidSpecificationException: ["ext/magic/share/magic.mgc"] are not files
/Users/stanhu/github/ruby-magic/rakefile:89:in `gem_builder'
/Users/stanhu/github/ruby-magic/rakefile:105:in `block (4 levels) in <top (required)>'
/usr/local/rvm/gems/ruby-2.5.8/gems/rake-13.0.3/exe/rake:27:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.5.8/bin/ruby_executable_hooks:24:in `eval'
/usr/local/rvm/gems/ruby-2.5.8/bin/ruby_executable_hooks:24:in `<main>'
Tasks: TOP => pkg/ruby-magic-0.4.0-x86_64-linux.gem
(See full trace by running task with --trace)
rake aborted!
Command failed with status (1): [docker run -v /Users/stanhu/github/ruby-magic:/Users/stanhu/github/ruby-magic -e UID\=1000 -e GID\=1000 -e USER\=stanhu -e GROUP\=_staff -e GEM_PRIVATE_KEY_PASSPHRASE -e ftp_proxy -e http_proxy -e https_proxy -e RCD_HOST_RUBY_PLATFORM\=x86_64-darwin19 -e RCD_HOST_RUBY_VERSION\=2.7.2 -e RCD_IMAGE\=larskanis/rake-compiler-dock-mri-x86_64-linux:1.1.0 -w /Users/stanhu/github/ruby-magic --rm -i -t larskanis/rake-compiler-dock-mri-x86_64-linux:1.1.0 runas sigfw bash -c gem\ install\ bundler\ --no-document\ \&\&'
'bundle\ \&\&'
'bundle\ exec\ rake\ gem:x86_64-linux:builder\ MAKE\=\'nice\ make\ -j\`nproc\`\''
']
/Users/stanhu/github/ruby-magic/Rakefile:95:in `block (3 levels) in <top (required)>'
/Users/stanhu/.rbenv/versions/2.7.2/bin/bundle:23:in `load'
/Users/stanhu/.rbenv/versions/2.7.2/bin/bundle:23:in `<main>'
Tasks: TOP => gem:x86_64-linux
(See full trace by running task with --trace)
```

Though the file is there locally:

```
$ pwd
/Users/stanhu/github/ruby-magic
$ ls ext/magic/share/magic.mgc
ext/magic/share/magic.mgc
```

flavorjones replied:
You may want to make sure you've blown away your `tmp` and `pkg` directories (my preference would be to have `tmp` in `CLEAN` rather than `CLOBBER` for this reason).

stanhu replied:
Ok, I ran a `rake clobber` and bumped my `VERSION` to 0.4.1 and got the `.gem` to build. After I installed it, I tried to use it:

```ruby
irb(main):001:0> require 'magic'
Traceback (most recent call last):
       10: from /Users/stanhu/.rbenv/versions/2.7.2/bin/irb:23:in `<main>'
        9: from /Users/stanhu/.rbenv/versions/2.7.2/bin/irb:23:in `load'
        8: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        7: from (irb):1
        6: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:156:in `require'
        5: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:168:in `rescue in require'
        4: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/2.7.0/rubygems/core_ext/kernel_require.rb:168:in `require'
        3: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic.rb:3:in `<top (required)>'
        2: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic.rb:8:in `rescue in <top (required)>'
        1: from /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic.rb:8:in `require_relative'
LoadError (cannot load such file -- /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic/magic)
```

Something isn't working here:

```ruby
begin
  ::RUBY_VERSION =~ /(\d+\.\d+)/
  require_relative "magic/#{Regexp.last_match(1)}/magic"
rescue LoadError => e
  require_relative 'magic/magic'
end
```

If I debug the exception in that `require` block, I see:

```
cannot load such file -- /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/lib/magic/2.7/magic
```

There's no `magic/2.7/magic` directory:

```
$ tree /Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/
/Users/stanhu/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/ruby-magic-0.4.1-x86_64-linux/
├── AUTHORS
├── CHANGELOG.md
├── CONTRIBUTORS.md
├── COPYRIGHT
├── LICENSE
├── NOTICE
├── README.md
├── dependencies.yml
├── ext
│   └── magic
│       ├── common.h
│       ├── extconf.rb
│       ├── functions.c
│       ├── functions.h
│       ├── ruby-magic.c
│       ├── ruby-magic.h
│       └── share
│           └── magic.mgc
├── kwilczynski-public.pem
└── lib
    ├── magic
    │   ├── 2.5
    │   │   └── magic.so
    │   ├── 2.6
    │   │   └── magic.so
    │   ├── 2.7
    │   │   └── magic.so
    │   ├── 3.0
    │   │   └── magic.so
    │   ├── core
    │   │   ├── file.rb
    │   │   └── string.rb
    │   └── version.rb
    └── magic.rb

10 directories, 24 files
```

I noticed if I renamed the `require_relative` calls to `require` above (as Nokogiri seems to do), it works?

flavorjones replied:
The reference to `lib/magic/2.7/magic` is to the file `lib/magic/2.7/magic.so` (not to a directory) which it looks like does exist on your system. I'm sorry, I don't fully understand what's going on on your system, and I'm having trouble reproducing it.

flavorjones replied:
Wait, are you trying to use the precompiled linux gem on MacOS?

![image](https://user-images.githubusercontent.com/8207/114392165-63e73b80-9b66-11eb-829f-6498f5f394e4.png)

stanhu replied:
🤦 Silly me! I was just testing you.

Re-tested this on a Linux system, and it worked fine. Thanks!

flavorjones commented:
@kwilczynski reported an issue with the current packaging at https://github.com/kwilczynski/ruby-magic/pull/16#issuecomment-817890141 which we should fix.

stanhu commented on ext/magic/ruby-magic.c:
```diff
@@ -294,15 +294,19 @@ rb_mgc_get_paths(VALUE object)
+		cstring = magic_getpath_wrapper();
+		value = magic_split(CSTR2RVAL(cstring), CSTR2RVAL(":"));
+	}
+	return magic_set_paths(object, value);
```
I don't know if it was coincidence, but my VM just stopped functioning after I ran this after installing the native package:

```shell
$ MAGIC=/tmp irb
irb(main):001:0> require 'magic'
=> true
irb(main):002:0> Magic.new
```

stanhu replied:
Not seeing this problem after a reboot. 🤷

stanhu commented:
@flavorjones Looks good, thanks!

kwilczynski commented:
Hi @flavorjones, thank you for working on this! Looks amazing!

I am seeing a similar problem to the one described by @stanhu in our CI, as per:

- https://travis-ci.com/github/kwilczynski/ruby-magic/jobs/498271374
- https://travis-ci.com/github/kwilczynski/ruby-magic/jobs/498271375

Not sure yet where the problem might be. I will look later.

Krzysztof
